### PR TITLE
[Train] Improvements to fault tolerance

### LINKED ISSUE #22349: [Bug] [Train] Cannot reproduce fault-tolerance, script hangs upon any node shutdown
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Train

### What happened + What you expected to happen

I just run the pytorch official Train+Torch DDP [here](https://github.com/ray-project/ray/blob/master/python/ray/train/examples/train_fashion_mnist_example.py) on a 2 node cluster on GCP (each with a single k80)
- It works, but whenever I try to kill one of them to test fault-tolerance, it hangs indefinitely waiting for another node
- Note that the cluster manager successfully launches a new node (as I have a min-workers of 2 in the script), but the script still hangs
- I tried having an additional available node (min-workers=3), but still the script hangs upon any node failure
- I feel it is a problem in matching requirements from Train with avail nodes
- Sometimes killing a single node, makes the whole training fails (not sure why)

### Versions / Dependencies

latest master & Pytorch 1.10

### Reproduction script

This is the cluster yaml

```
cluster_name: my-cluster

upscaling_speed: 10.0
idle_timeout_minutes: 50

min_workers: 4
max_workers: 40

provider:
    type: gcp
    region: us-central1
    availability_zone: us-central1-a
    project_id: my-proj

auth:
    ssh_user: user

available_node_types:
    ray_head_default:
        min_workers: 0
        max_workers: 0
        resources: {"CPU": 0}
        node_config:
            machineType: n1-standard-2
            disks:
              - boot: true
                autoDelete: true
                type: PERSISTENT
                initializeParams:
                  diskSizeGb: 100
                  sourceImage: projects/path/to/myimage
            scheduling:
              - provisioningModel: SPOT
              - onHostMaintenance: TERMINATE
            
    ray_worker_small:
        min_workers: 2
        resources: {"CPU": 8, "GPU": 1.0}
        node_config:
            machineType: n1-standard-8
            disks:
              - boot: true
                autoDelete: true
                type: PERSISTENT
                initializeParams:
                  diskSizeGb: 100
                  sourceImage: projects/path/to/myimage

            guestAccelerators:
              - acceleratorType: nvidia-tesla-k80
                acceleratorCount: 1
            scheduling:
              - provisioningModel: SPOT
              - onHostMaintenance: TERMINATE
            
            networkInterfaces:
              - accessConfigs:
                - name: "External NAT"
                nicType: GVNIC
                subnetwork: "projects/my-proj/regions/us-central1/subnetworks/default"

setup_commands:
    - pip uninstall -y ray
    - pip install -U https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-2.0.0.dev0-cp37-cp37m-manylinux2014_x86_64.whl
```

### Anything else

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by amogkam**
Various improvements to Ray Train fault tolerance.

- Add more log statements for better debugging of Ray Train failure handling.
- Fixes https://github.com/ray-project/ray/issues/22349. ~~The default torch process group timeout is 30 minutes. If a failure occurs before a gradient synchronization, training will hang for 30 minutes before raising an error and triggering fault tolerance. This PR reduces the default timeout_s to 30 seconds.~~ 
~~Adds functionality to trigger fault tolerance even if any of the alive workers are hanging by specifying a grace period for alive workers.~~
Simplifies fault tolerance by removing backend specific `handle_failure`. If any workers have failed, all workers will be restarted and training will continue from the last checkpoint.
Also adds a test for fault tolerance with an actual torch example. When testing locally, the test hangs before the fix, but passes after. 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/backend.py [hunk 1]
```diff
@@ -67,8 +67,11 @@ def handle_failure(
 
         By default, restart all workers.
         """
+        logger.info("Shutting down all training workers.")
         worker_group.shutdown()
+        logger.info("Restarting all training workers.")
         worker_group.start()
+        logger.info("Setting up distributed backend on restarted workers.")
```
[the remark this answers is not in the document]
Comment (amogkam) [reply python/ray/train/backend.py:74]: Fixed

File: python/ray/train/trainer.py [hunk 2]
```diff
@@ -720,6 +720,13 @@ def _run_with_error_handling(self, func: Callable):
             return func()
         except TrainingWorkerError:
             # Workers have already been restarted.
+            logger.info(
+                "Workers have been successfully restarted. Resuming "
+                "training from latest checkpoint."
+            )
+            logger.debug(
+                "Latest checkpoint:" f" {self._checkpoint_manager.latest_checkpoint}"
```
[the remark this answers is not in the document]
Comment (amogkam) [reply python/ray/train/trainer.py]: Updated! ...blame the linter

File: python/ray/train/backend.py [hunk 3]
```diff
@@ -67,8 +67,11 @@ def handle_failure(
 
         By default, restart all workers.
         """
+        logger.info("Shutting down all training workers.")
         worker_group.shutdown()
+        logger.info("Restarting all training workers.")
         worker_group.start()
+        logger.info("Setting up distributed backend on all workers.")
```
Comment (jovany-wang) [inline python/ray/train/backend.py:74]: Are these logs too verbose?

File: python/ray/train/utils.py [hunk 4]
```diff
@@ -31,19 +33,46 @@ def check_for_failure(remote_values: List[ObjectRef]) -> Tuple[bool, List[int]]:
...
     dead_worker_indexes = []  # Store the indexes of the failed workers.
+
+    at_least_one_failed_worker = False
     while len(unfinished) > 0:
-        finished, unfinished = ray.wait(unfinished)
+        # Once a worker has failed, we add a timeout for getting the
+        # results for the remaining workers.
+        # This is to avoid situations where the remaining workers
+        # are alive, but hanging on collective calls because other
+        # workers have failed.
```
Comment (matthewdeng) [inline python/ray/train/utils.py:43]: Can this be generalized? Is this exclusive to collective calls or are there other situations this would help as well?

File: python/ray/train/utils.py [hunk 5]
```diff
@@ -31,19 +33,46 @@ def check_for_failure(remote_values: List[ObjectRef]) -> Tuple[bool, List[int]]:
...
+        timeout = (
+            REMAINING_WORKERS_GRACE_PERIOD_S if at_least_one_failed_worker else None
+        )
+        finished, unfinished = ray.wait(unfinished, timeout=timeout)
+
+        if at_least_one_failed_worker and len(unfinished) > 0:
+            # If at least one worker has failed, but there are still workers
+            # that we are waiting on results from, these workers are hanging.
+            # Treat these workers as dead workers as well.
```
Comment (matthewdeng) [inline python/ray/train/utils.py:52]: Do we actually need to treat these as dead? Or is it sufficient that we break out of this `ray.wait` call and only return the failed worker as dead?

File: python/ray/train/utils.py [hunk 6]
```diff
@@ -31,19 +33,46 @@ def check_for_failure(remote_values: List[ObjectRef]) -> Tuple[bool, List[int]]:
...
+            unfinished_indexes = [remote_values.index(i) for i in unfinished]
+            dead_worker_indexes.extend(unfinished_indexes)
+            unfinished = []
+
         # If a failure occurs the ObjectRef will be marked as finished.
         # Calling ray.get will expose the failure as a RayActorError.
         for object_ref in finished:
+            # Everything in finished has either failed or
```
Comment (matthewdeng) [inline python/ray/train/utils.py:60]: or what 😨

File: python/ray/train/utils.py [hunk 7]
```diff
@@ -31,19 +33,46 @@ def check_for_failure(remote_values: List[ObjectRef]) -> Tuple[bool, List[int]]:
...
+    at_least_one_failed_worker = False
     while len(unfinished) > 0:
-        finished, unfinished = ray.wait(unfinished)
+        # Once a worker has failed, we add a timeout for getting the
+        # results for the remaining workers.
+        # This is to avoid situations where the remaining workers
+        # are alive, but hanging on collective calls because other
+        # workers have failed.
+        timeout = (
+            REMAINING_WORKERS_GRACE_PERIOD_S if at_least_one_failed_worker else None
```
Comment (matthewdeng) [inline python/ray/train/utils.py:45]: Change `REMAINING_WORKERS_GRACE_PERIOD_S` to support being set as an environment variable? Is 10 seconds too short?

File: python/ray/train/utils.py (hunk 4 above)
Comment (amogkam) [reply python/ray/train/utils.py:43]: Any situation where some workers dying cause other workers to hang.

Comment (amogkam) [reply python/ray/train/utils.py:43]: Updated the comment to reflect this

File: python/ray/train/utils.py (hunk 5 above)
Comment (amogkam) [reply python/ray/train/utils.py:52]: That's a good point, we don't need to treat them as dead. Will make the change.

File: python/ray/train/utils.py (hunk 6 above)
Comment (amogkam) [reply python/ray/train/utils.py:60]: Or completed successfully 😁...good catch, updated!

File: python/ray/train/backend.py (hunk 3 above)
Comment (amogkam) [reply python/ray/train/backend.py:74]: Hmmm, most people don't run with log level DEBUG, and for very long running training jobs (15+ hours) I think these messages could be useful and it's not feasible to run the training job again.

cc @matthewdeng would also like to hear your thoughts here.

File: python/ray/train/utils.py (hunk 5 above)
Comment (amogkam) [reply python/ray/train/utils.py:52]: Actually no, these have to be marked as dead because new actors may not be created. And these actors are no longer usable since they are hanging on a particular method execution.

File: python/ray/train/backend.py (hunk 3 above)
Comment (matthewdeng) [reply python/ray/train/backend.py:74]: I think INFO is fine given that this is only logged during failures, which should hopefully be rare.

File: python/ray/train/utils.py [hunk 8]
```diff
@@ -33,33 +33,35 @@
...
-def check_for_failure(remote_values: List[ObjectRef]) -> Tuple[bool, List[int]]:
+def check_for_failure(remote_values: List[ObjectRef]) -> bool:
     """Check for actor failure when retrieving the remote values.
 
     Args:
         remote_values (list): List of object references from Ray actor methods.
 
     Returns:
-        Returns Tuple of success boolean and list of workers indexes that fail.
+        True evaluating all object references is successful, False otherwise.
```
Comment (amogkam) [inline python/ray/train/utils.py:43]:
```suggestion
        True if evaluating all object references is successful, False otherwise.
```

**Review by matthewdeng — APPROVED** [review]
LGTM - I think you just need to pull in master to fix the lint issues.

File: python/ray/train/utils.py [hunk 9]
```diff
@@ -33,33 +33,35 @@
...
             except RayActorError as exc:
                 logger.exception(str(exc))
                 failed_actor_rank = remote_values.index(object_ref)
                 logger.info(f"Worker {failed_actor_rank} has failed.")
-                dead_worker_indexes.append(failed_actor_rank)
-    if len(dead_worker_indexes) > 0:
-        return False, dead_worker_indexes
-    else:
-        return True, []
+                at_least_one_failed_worker = True
```
Comment (matthewdeng) [inline python/ray/train/utils.py:62]: nit: Just return `False` here? Or remove `and not at_least_one_failed_worker` in line 48 if you want to print all failed workers.

Comment (amogkam) [reply python/ray/train/utils.py:62]: good point, updated!
